bash_completion

```diff
@@ -149,6 +149,22 @@ dequote()
     eval printf %s "$1" 2>/dev/null
 }
 
+# Unset the given variables across a scope boundary. Useful for unshadowing
+# global scoped variables. Note that simply calling unset on a local variable
+# will not unshadow the global variable. Rather, the result will be a local
+# variable in an unset state.
+# Usage: local IFS=$'|'; _comp_unlocal IFS
+# Param: $* Variable names to be unset
+_comp_unlocal()
+{
+    local reset=$([[ ${BASH_VERSINFO[0]} -ge 5 ]] && {
+                      shopt -p localvar_unset
+                      shopt -u localvar_unset
+                  })
+    unset -v "$@"
+    $reset
```

Comment:
I think we here assume `IFS=$' \t\n'` or unset `IFS` to make `$reset` work, but we don't know the actual value of the global `IFS`. Can you use a different approach? Also, I actually don't like the subshell `$()` which has the fork cost.

Actually, there are the same problems in other places, which should be resolved in the future but that doesn't need to be addressed in this PR. Maybe we can open a discussion about what would be a better replacement for the `reset=$(shopt -p)` / `$reset` idiom currently used in many places.

Reply:
Perhaps something leveraging a map, and `shopt -q` could be a general solution here. Let me go try something.

Reply:
```
store_options() {
    declare -n state=$1
    shift
    for opt in $@
    do
        shopt -q $opt && state[$opt]="-s" || state[$opt]="-u"
    done
}

restore_options() {
    declare -n state=$1
    shift
    for opt in ${!state[@]}
    do
        shopt ${state[$opt]} $opt
    done
}

some_function() {
    declare -A opts
    set -x
    store_options opts localvar_unset

    restore_options opts
    set +x
}
```
What about this?

Reply:
Maybe we can just save the value of `$BASHOPTS`, and then later test it using `[[ :$saved_bashopts: == *:"$opt":* ]]`. This is actually what I do in my personal scripts.


```bash
_comp_restore_shopt() {
    local bashopts=$1; shift
    local opt
    for opt; do
        if [[ :$bashopts: == *:"$opt":* ]]; then
            shopt -s "$opt"
        else
            shopt -u "$opt"
        fi
    done
}
some_function() {
    local bashopts=$BASHOPTS
    set -x
    shopt -u localvar_unset

    _comp_restore_shopt "$bashopts" localvar_unset
    set +x
}
```

Reply:
I like this approach overall, and it is straightforward. The only concern I have is that it is easy to forget an argument when calling `_comp_restore_shopt`. That could be avoided by looping over all options like this:

```
_comp_restore_shopt() {
    local bashopts=$1; shift
    local IFS=:
    for opt in $BASHOPTS $bashopts; do
        if [[ :$bashopts: == *:"$opt":* ]]; then
            shopt -s "$opt"
        else
            shopt -u "$opt"
        fi
    done
}
```

I think this approach is likely not worth the loss of efficiency compared with your approach.

Reply:
I have created an issue at #691.